application/comit_node/src/connection_pool.rs

```diff
@@ -0,0 +1,286 @@
+                    Err(e)
+                }
+            },
+            // Somehow we got into Disconnected while connecting
+            // Just transition to Connected. (I guess?)
+            ConnectionState::Disconnected => match socket {
+                Ok(socket) => Ok(Self::transition_to_connected(
+                    node_id,
+                    &mut *connection_state,
+                    Arc::clone(&connection_handle),
+                    bob_spawner,
+                    socket,
+                )),
+                Err(e) => {
+                    error!("Failed to connect to {}: {:?}", node_id, e);
+                    Err(e)
+                }
+            },
+        }
+    }
+
+    /// Takes a server socket and updates the
+    pub fn new_incoming_socket<B: BobSpawner>(&self, socket: TcpStream, bob_spawner: B) {
+        let node_id = match socket.peer_addr() {
+            Ok(node_id) => node_id,
+            Err(e) => {
+                error!("Couldn't get peer address: {:?}", e);
+                return;
+            }
+        };
+        let mut connections = self.connections.write().unwrap();
+
+        let existing_connection = connections
+            .entry(node_id)
+            .or_insert_with(|| Arc::new(Mutex::new(ConnectionState::Disconnected)));
+
+        let mut existing_connection_guard = (*existing_connection).lock().unwrap();
+
+        match *existing_connection_guard {
+            // Ignore the new socket -- we already have a connection
+            ConnectionState::Connected { .. } => {
+                warn!("Ignoring incoming connection from {} because we already have a connection to them", node_id);
+            }
+
+            // We're trying to connect to the remote node at the same time they
+            // connect to us. We ignore their connection and keep connecting.
+            ConnectionState::Connecting { .. } => {
+                warn!(
+                    "Ignoring incoming connection from {} because we're already connecting",
+                    node_id
+                );
+            }
+            // The most common case. We were disconnected but now we're connected.
+            ConnectionState::Disconnected => {
+                Self::transition_to_connected(
+                    node_id,
+                    &mut *existing_connection_guard,
+                    Arc::clone(&existing_connection),
+                    bob_spawner,
+                    socket,
+                );
+            }
+        }
+    }
+
+    /// Sets the connection_state to "Connected" and spawns the Connection.
+    /// When the connection is finished it sets it to Disconnected.
+    /// `connection_state` and `connection_handle`` must point to the same
+    /// ConnectionState.
+    fn transition_to_connected<S: AsyncRead + AsyncWrite + Send + 'static, B: BobSpawner>(
+        node_id: NodeId,
+        connection_state: &mut ConnectionState,
```

Comment:
:question: Is this the `current` ConnectionState? Maybe distinct naming could avoid this confusion.

Reply:
Yes. I thought that "current" is implied